### PR TITLE
[SPARK-23528][ML] Add numIter to ClusteringSummary

### PR DESCRIPTION
**Pull request by mgaido91**
## What changes were proposed in this pull request?

Added the number of iterations in `ClusteringSummary`. This is an helpful information in evaluating how to eventually modify the parameters in order to get a better model.

## How was this patch tested?

modified existing UTs

### CODE REVIEW

**Comment by mgaido91** [comment]
cc @yanboliang @zhengruifeng since I saw you worked on this before, thanks.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeansModel.scala
Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeansModel.scala]: So were using -1 to indicate we don't have the numIter information

File: mllib/src/test/scala/org/apache/spark/ml/clustering/BisectingKMeansSuite.scala
Comment (holdenk) [inline mllib/src/test/scala/org/apache/spark/ml/clustering/BisectingKMeansSuite.scala]: So I'd be more comfortable having this in a separate test, 2 iterations is not a lot.

File: mllib/src/test/scala/org/apache/spark/ml/clustering/BisectingKMeansSuite.scala
Comment (holdenk) [inline mllib/src/test/scala/org/apache/spark/ml/clustering/BisectingKMeansSuite.scala]: Would be nice to see a test where its not maxIter value being copied over

**Review by holdenk — COMMENTED** [review]
I've done a quick pass and I'm going to see if @sethah has some comments.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeansModel.scala
Comment (mgaido91) [reply mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeansModel.scala]: yes, this can happen for instance when reloading a persisted model. Moreover this is only for the mllib model, which as far as I know is suggested not to be used anymore in favor of the new ml api. Any concern/suggestion about this?

File: mllib/src/test/scala/org/apache/spark/ml/clustering/BisectingKMeansSuite.scala
Comment (mgaido91) [reply mllib/src/test/scala/org/apache/spark/ml/clustering/BisectingKMeansSuite.scala]: In `KMeansSuite` the value is not `maxIter` (it performs only 1 iteration in that case). In `BisectingKMeans` `numIter` is always `maxIter` since we are always performing `maxIter` (see https://github.com/apache/spark/blob/b6f837c9d3cb0f76f0a52df37e34aea8944f6867/mllib/src/main/scala/org/apache/spark/mllib/clustering/BisectingKMeans.scala#L192).

Does it answer to your comment?

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeansModel.scala
Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeansModel.scala]: So I think the correct since annotation here would be 0.8.0 since this is just a move of the previous constructor right?

File: mllib/src/main/scala/org/apache/spark/ml/clustering/BisectingKMeans.scala
Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/BisectingKMeans.scala]: Here (and in the others), we should add this as param in the comment above as done with the other params

File: mllib/src/main/scala/org/apache/spark/ml/clustering/ClusteringSummary.scala
Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/ClusteringSummary.scala]: Please add this param in the comment above.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeansModel.scala
Comment (holdenk) [reply mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeansModel.scala]: Sounds reasonable, I personally don't enjoy -1 to indicate lack of information but it seems to be what we have generally used in the past for mllib summary info into ml so my personal feelings aren't important :)

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeansModel.scala
Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeansModel.scala]: So previously the main constructor was not private, any particular reason we are making in private? if someone else is implementing something which extends the kmeans model this might be a little frustrating.

File: project/MimaExcludes.scala
Comment (holdenk) [inline project/MimaExcludes.scala]: Just a note for other reviewers/myself these are all private spark constructors

**Review by holdenk — CHANGES_REQUESTED** [review]
Thanks for working on this, good progress. I have a few improvements in mind, and maybe we can get @sethah to take a look as well, but if the rest of the ML committers are busy thats ok too.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeansModel.scala
Comment (mgaido91) [reply mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeansModel.scala]: I think this is the right one. 0.8.0 is the annotation for the `KMeansModel` class, while the previous main constructor was added (by me) is a previous PR for 2.4.0 in order to add the `distanceMeasure` variable.

File: mllib/src/main/scala/org/apache/spark/ml/clustering/BisectingKMeans.scala
Comment (mgaido91) [reply mllib/src/main/scala/org/apache/spark/ml/clustering/BisectingKMeans.scala]: thanks for pointing this out, I completely missed it. Thank you, I am adding them.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeansModel.scala
Comment (mgaido91) [reply mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeansModel.scala]: I just didn't want the user to be able to create a KMeansModel setting the number of iterations. I moved the other constructor which is still available. I don't have strong reasons against making this public, so I am removing the private clause if you think we best let it to be public.

**Comment by mgaido91** [comment]
retest this please

**Comment by mgaido91** [comment]
retest this please

**Comment by mgaido91** [comment]
any more comments @holdenk ?

**Review by sethah — COMMENTED** [review]
General comment: things that are specific to training, like numIter, have been separated into training summary classes elsewhere, e.g. `LinearRegressionTrainingSummary extends LinearRegressionSummary`. Is there some reason to deviate from that here? `numIter` doesn't make sense when evaluating on a test set, for instance.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeansModel.scala
Comment (mgaido91) [reply mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeansModel.scala]: yes, I will make it private, thanks.

**Comment by mgaido91** [comment]
@sethah I have not introduces training summary classes because it would have meant a quite bigger change - since they have a quite different approach, having a trait and an Impl class for each of them - and I have not seen that pattern to be always used.

**Comment by mgaido91** [comment]
retest this please

**Comment by mgaido91** [comment]
any more comments @holdenk @sethah ?

**Comment by holdenk** [comment]
ping @sethah - what do you think about if this needs a separate training summary trait?

**Comment by holdenk** [comment]
So we need to update for the changed MimaExcludes, I think its ok to include this in the model directly if no one objects in the next week or so? Sklearn has this directly in the model return as well. Ping @sethah @MLnick .

**Comment by mgaido91** [comment]
kindly ping @holdenk

**Comment by mgaido91** [comment]
kindly ping @holdenk

**Comment by holdenk** [comment]
LGTM pending Jenkins retest. Jenkins retest this please.

**Comment by holdenk** [comment]
The AppVeyor build failure looks spurrious and I don't know how to retrigger it.

**Comment by mgaido91** [comment]
retest this please

**Comment by mgaido91** [comment]
thanks for your review @holdenk. I don't know how to retrigger AppVeyor too, unfortunately :(

**Review by holdenk — APPROVED** [review]
Looks good to me, manually built docs locally to double check the since annotation was inherited because my memory was a bit fuzzy on how that was handled.

**Pull request closed without merging**
